src/framework/audio/internal/sequencer.cpp

```diff
@@ -0,0 +1,291 @@
+    }
+
+    if (m_nextSeek.has_value()) {
+        auto miliseconds = m_nextSeek.value_or(0);
+        m_nextSeek.reset();
+
+        m_clock->seekMiliseconds(miliseconds);
+        for (auto& track : m_tracks) {
+            track.second->seek(miliseconds);
+        }
+        m_positionChanged.notify();
+    }
+
+    std::for_each(m_backgroudPlayers.begin(), m_backgroudPlayers.end(), [&miliseconds](auto& player) {
+        player.first += miliseconds;
+        player.second->forwardTime(player.first);
+    });
+
+    m_backgroudPlayers.remove_if([](auto player) {
+        return player.second->status() == IPlayer::Status::Stoped;
+    });
+}
+
+void Sequencer::setStatus(Status status)
+{
+    m_status = status;
+    m_statusChanged.send(m_status);
+
+    std::function<void(const track_t&)> applyStaus = [](const track_t&) {};
+    switch (m_status) {
+    case PLAYING:
+        applyStaus = [](const track_t& track) { track->play(); };
+        m_clock->start();
+        break;
+    case STOPED:
+        applyStaus = [](const track_t& track) { track->stop(); };
+        m_clock->stop();
+        break;
+    }
+
+    for (auto& track : m_tracks) {
+        applyStaus(track.second);
+    }
+}
+
+std::optional<Sequencer::track_t> Sequencer::track(Sequencer::track_id id) const
+{
+    if (m_tracks.find(id) != m_tracks.end()) {
+        return m_tracks.at(id);
+    }
+    return std::nullopt;
+}
+
+Sequencer::midi_track_t Sequencer::midiTrack(Sequencer::track_id id) const
+{
+    if (auto track_var = track(id)) {
+        if (auto track_value = track_var.value_or(nullptr)) {
+            return std::dynamic_pointer_cast<IMIDIPlayer>(track_value);
+        }
+    }
+    return nullptr;
+}
+
+Sequencer::midi_track_t Sequencer::createMIDITrack(ISequencer::track_id id)
+{
+    auto player = std::make_shared<MIDIPlayer>();
+    m_tracks[id] = player;
+    return player;
+}
+
+Sequencer::audio_track_t Sequencer::createAudioTrack(Sequencer::track_id id)
+{
+    auto player = std::make_shared<AudioPlayer>();
+    m_tracks[id] = player;
+    m_audioTrackAdded.send(player);
+    return player;
+}
+
+Sequencer::audio_track_t Sequencer::audioTrack(Sequencer::track_id id) const
+{
+    if (auto track_var = track(id)) {
+        if (auto track_value = track_var.value_or(nullptr)) {
+            return std::dynamic_pointer_cast<IAudioPlayer>(track_value);
+        }
+    }
+    return nullptr;
+}
+
+std::shared_ptr<IMIDIPlayer> Sequencer::instantlyPlayMidi(const midi::MidiData& data)
+{
+    if (!data.isValid()) {
+        return nullptr;
+    }
+    auto player = std::make_shared<MIDIPlayer>();
+    auto midiStream = std::make_shared<midi::MidiStream>();
+    std::lock_guard guard(m_syncMutex);
+
+    midiStream->initData = data;
+    midiStream->isStreamingAllowed = false;
+    midiStream->lastTick = data.lastChunksTick();
+    player->loadMIDI(midiStream);
+    player->play();
+
+    m_backgroudPlayers.push_back({ 0 /*ms*/, player });
```

Comment:
It requires additional research why these players cannot be regular players (putting into m_tracks).